packages/core/src/stateUtils.ts

```diff
@@ -1391,22 +1389,40 @@ function selectEventlessTransitions<TContext, TEvent extends EventObject>(
   state: State<TContext, TEvent>,
   machine: MachineNode<TContext, any, TEvent>
 ): Transitions<TContext, TEvent> {
-  const transientNodes = state.configuration.filter(sn => sn.isTransient);
-
-  const transitions = flatten(transientNodes.map(sn => sn.transitions));
-  return transitions.filter(t => {
-    return (
-      t.eventType === NULL_EVENT &&
-      (t.cond === undefined ||
-        evaluateGuard<TContext, TEvent>(
-          machine,
-          t.cond,
-          state.context,
-          toSCXMLEvent(NULL_EVENT as Event<TEvent>),
-          state
-        ))
-    );
-  });
+  const enabledTransitions: Set<
+    TransitionDefinition<TContext, TEvent>
+  > = new Set();
+
+  const atomicStates = state.configuration.filter(isAtomicStateNode);
+
+  for (const stateNode of atomicStates) {
+    loop: for (const s of [stateNode].concat(
```

Comment:
it's been a while since I've seen a label 😄

Reply:
Comes straight from the algorithm! 😅